### PR TITLE
feat: skip background refresh for unchanged insights

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

The more dashboards and insights we have, the more we try to update them in the background, the longer it takes us to refresh them

We refresh the insight even if the data hasn't changed.

## Changes

Uses the EventDefinition "last seen at" and the insight/tile last refresh date to guess if refresh can be skipped.

If the events for an Insight have been ingested since the last refresh then the data will (possibly) have changed and we should refresh the insight.

If the events have not been ingested then the existing cached results are still valid and we don't need to spend time refreshing the insight. 

## How did you test this code?

Adding developer tests and running it locally with breakpoints to show that the update uses both the old and new path through the code.

### CODE REVIEW

File: posthog/tasks/update_cache.py [hunk 1]
```diff
@@ -203,64 +274,91 @@ def update_cache_item(key: str, cache_type: CacheType, payload: dict) -> List[Di
...
+    and last_refresh is after last_seen_at for each event in the filter
+
+    then there's no point re-calculating
+    """
+
+    # dashboards mostly use trends and funnels so concentrate on items that have filter.events to interrogate
+
+    last_refresh = payload.get("last_refresh", None)
+    if last_refresh:
+        event_ids = _events_from_filter(filter)
```
Comment (Twixes) [inline posthog/tasks/update_cache.py:333]: `event_ids` made me think of `Event.id`, when this is `Event.event`, so maybe `event_names`?

File: posthog/tasks/update_cache.py [hunk 2]
```diff
@@ -203,64 +274,91 @@ def update_cache_item(key: str, cache_type: CacheType, payload: dict) -> List[Di
...
+def _events_from_filter(filter: Union[RetentionFilter, StickinessFilter, PathFilter, Filter]) -> List[str]:
+    try:
+        if isinstance(filter, RetentionFilter):
+            event_ids = [
+                str(filter.target_entity.id),
+                str(filter.returning_entity.id),
+            ]
+
+        elif isinstance(filter, PathFilter):
+            event_ids = filter.target_events
```
Comment (Twixes) [inline posthog/tasks/update_cache.py:353]: `target_events` isn't exactly event names, it's basically a `Set["$pageview" | "$screen" | "custom_event"]`. That last variant is a shorthand for all events other than `$pageview`/`$screen`.

File: posthog/tasks/update_cache.py [hunk 3]
```diff
@@ -203,64 +274,91 @@ def update_cache_item(key: str, cache_type: CacheType, payload: dict) -> List[Di
...
+
+        else:
+            event_ids = [str(e.id) for e in filter.events]
```
Comment (Twixes) [inline posthog/tasks/update_cache.py:356]: This ignores series based on an action, which use event definitions indirectly. Also, actions can change at any time. So we should still always refresh insights based on actions.

File: posthog/tasks/update_cache.py [hunk 4]
```diff
@@ -203,64 +274,91 @@ def update_cache_item(key: str, cache_type: CacheType, payload: dict) -> List[Di
...
+
+    return False
+
+
+def _events_from_filter(filter: Union[RetentionFilter, StickinessFilter, PathFilter, Filter]) -> List[str]:
+    try:
+        if isinstance(filter, RetentionFilter):
+            event_ids = [
+                str(filter.target_entity.id),
+                str(filter.returning_entity.id),
```
Comment (Twixes) [inline posthog/tasks/update_cache.py:349]: What if those are actions?

**Review by Twixes — CHANGES_REQUESTED** [review]

File: posthog/tasks/update_cache.py (hunk 2 above)
Comment (pauldambra) [reply posthog/tasks/update_cache.py:353]: Trends are the bulk of the cache time (I believe) so to keep this simple so we can test it I'll not try to paths and retention types

File: posthog/tasks/update_cache.py (hunk 4 above)
Comment (pauldambra) [reply posthog/tasks/update_cache.py:349]: Trends are the bulk of the cache time (I believe) so to keep this simple so we can test it I'll not try to paths and retention types

File: posthog/tasks/update_cache.py [hunk 5]
```diff
@@ -330,32 +330,31 @@ def _cache_includes_latest_events(
...
+    so for now we'll just return an empty list and can (dis?)prove that this mechanism is useful
+    """
+    try:
+        if isinstance(filter, StickinessFilter) or isinstance(filter, Filter):
+            if not filter.actions:
+                return [str(e.id) for e in filter.events]
 
-        return event_ids
+        return []
     except Exception as exc:
```
Comment (Twixes) [inline posthog/tasks/update_cache.py:358]: Will this be logged in Sentry?

**Review by Twixes — APPROVED** [review]

**Pull request merged**